ddtrace/utils/cache.py

```diff
@@ -0,0 +1,81 @@
+from threading import RLock
+
+
+def cached(maxsize=256):
```

Comment:
this feels maybe like overkill for something we could preprocess once?

for example this is where we set the `config.http_server.error_statuses` config value:

https://github.com/DataDog/dd-trace-py/blob/4c47b26cba044cb5e916b0d11d2cfcb9518f7a00/ddtrace/settings/config.py#L63


But maybe we just have something like:

```python
class HTTPServerConfig:
    _error_statuses = []

    @property
    def error_statuses(self):
        return self._error_statuses

    @error_statuses.setter
    def error_statuses(self, value):
        self._error_statuses = parse_error_statuses(value)
```

Reply:
Agreed that `get_error_ranges` could be done at initialization but we could cache `is_error_code'.

Reply:
if we parse at initialization, not sure we need to cache this? we _could_ but seems like the big overhead is the parsing not the looking through ranges ?

Reply:
... we could pre-generate/cache all error status values.

since there is a finite set of status codes, shouldn't be much to store in memory:

```python
>>> headers = dict([(code, 500 <= code <= 599) for code in range(100, 599)])
>>> headers
{100: False, 101: False, 102: False, 103: False, 104: False, 105: False, 106: False, 107: False, 108: False, 109: False, 110: False, 111: False, 112: False, 113: False, 114: False, 115: False, 116: False, 117: False, 118: False, 119: False, 120: False, 121: False, 122: False, 123: False, 124: False, 125: False, 126: False, 127: False, 128: False, 129: False, 130: False, 131: False, 132: False, 133: False, 134: False, 135: False, 136: False, 137: False, 138: False, 139: False, 140: False, 141: False, 142: False, 143: False, 144: False, 145: False, 146: False, 147: False, 148: False, 149: False, 150: False, 151: False, 152: False, 153: False, 154: False, 155: False, 156: False, 157: False, 158: False, 159: False, 160: False, 161: False, 162: False, 163: False, 164: False, 165: False, 166: False, 167: False, 168: False, 169: False, 170: False, 171: False, 172: False, 173: False, 174: False, 175: False, 176: False, 177: False, 178: False, 179: False, 180: False, 181: False, 182: False, 183: False, 184: False, 185: False, 186: False, 187: False, 188: False, 189: False, 190: False, 191: False, 192: False, 193: False, 194: False, 195: False, 196: False, 197: False, 198: False, 199: False, 200: False, 201: False, 202: False, 203: False, 204: False, 205: False, 206: False, 207: False, 208: False, 209: False, 210: False, 211: False, 212: False, 213: False, 214: False, 215: False, 216: False, 217: False, 218: False, 219: False, 220: False, 221: False, 222: False, 223: False, 224: False, 225: False, 226: False, 227: False, 228: False, 229: False, 230: False, 231: False, 232: False, 233: False, 234: False, 235: False, 236: False, 237: False, 238: False, 239: False, 240: False, 241: False, 242: False, 243: False, 244: False, 245: False, 246: False, 247: False, 248: False, 249: False, 250: False, 251: False, 252: False, 253: False, 254: False, 255: False, 256: False, 257: False, 258: False, 259: False, 260: False, 261: False, 262: False, 263: False, 264: False, 265: False, 266: False, 267: False, 268: False, 269: False, 270: False, 271: False, 272: False, 273: False, 274: False, 275: False, 276: False, 277: False, 278: False, 279: False, 280: False, 281: False, 282: False, 283: False, 284: False, 285: False, 286: False, 287: False, 288: False, 289: False, 290: False, 291: False, 292: False, 293: False, 294: False, 295: False, 296: False, 297: False, 298: False, 299: False, 300: False, 301: False, 302: False, 303: False, 304: False, 305: False, 306: False, 307: False, 308: False, 309: False, 310: False, 311: False, 312: False, 313: False, 314: False, 315: False, 316: False, 317: False, 318: False, 319: False, 320: False, 321: False, 322: False, 323: False, 324: False, 325: False, 326: False, 327: False, 328: False, 329: False, 330: False, 331: False, 332: False, 333: False, 334: False, 335: False, 336: False, 337: False, 338: False, 339: False, 340: False, 341: False, 342: False, 343: False, 344: False, 345: False, 346: False, 347: False, 348: False, 349: False, 350: False, 351: False, 352: False, 353: False, 354: False, 355: False, 356: False, 357: False, 358: False, 359: False, 360: False, 361: False, 362: False, 363: False, 364: False, 365: False, 366: False, 367: False, 368: False, 369: False, 370: False, 371: False, 372: False, 373: False, 374: False, 375: False, 376: False, 377: False, 378: False, 379: False, 380: False, 381: False, 382: False, 383: False, 384: False, 385: False, 386: False, 387: False, 388: False, 389: False, 390: False, 391: False, 392: False, 393: False, 394: False, 395: False, 396: False, 397: False, 398: False, 399: False, 400: False, 401: False, 402: False, 403: False, 404: False, 405: False, 406: False, 407: False, 408: False, 409: False, 410: False, 411: False, 412: False, 413: False, 414: False, 415: False, 416: False, 417: False, 418: False, 419: False, 420: False, 421: False, 422: False, 423: False, 424: False, 425: False, 426: False, 427: False, 428: False, 429: False, 430: False, 431: False, 432: False, 433: False, 434: False, 435: False, 436: False, 437: False, 438: False, 439: False, 440: False, 441: False, 442: False, 443: False, 444: False, 445: False, 446: False, 447: False, 448: False, 449: False, 450: False, 451: False, 452: False, 453: False, 454: False, 455: False, 456: False, 457: False, 458: False, 459: False, 460: False, 461: False, 462: False, 463: False, 464: False, 465: False, 466: False, 467: False, 468: False, 469: False, 470: False, 471: False, 472: False, 473: False, 474: False, 475: False, 476: False, 477: False, 478: False, 479: False, 480: False, 481: False, 482: False, 483: False, 484: False, 485: False, 486: False, 487: False, 488: False, 489: False, 490: False, 491: False, 492: False, 493: False, 494: False, 495: False, 496: False, 497: False, 498: False, 499: False, 500: True, 501: True, 502: True, 503: True, 504: True, 505: True, 506: True, 507: True, 508: True, 509: True, 510: True, 511: True, 512: True, 513: True, 514: True, 515: True, 516: True, 517: True, 518: True, 519: True, 520: True, 521: True, 522: True, 523: True, 524: True, 525: True, 526: True, 527: True, 528: True, 529: True, 530: True, 531: True, 532: True, 533: True, 534: True, 535: True, 536: True, 537: True, 538: True, 539: True, 540: True, 541: True, 542: True, 543: True, 544: True, 545: True, 546: True, 547: True, 548: True, 549: True, 550: True, 551: True, 552: True, 553: True, 554: True, 555: True, 556: True, 557: True, 558: True, 559: True, 560: True, 561: True, 562: True, 563: True, 564: True, 565: True, 566: True, 567: True, 568: True, 569: True, 570: True, 571: True, 572: True, 573: True, 574: True, 575: True, 576: True, 577: True, 578: True, 579: True, 580: True, 581: True, 582: True, 583: True, 584: True, 585: True, 586: True, 587: True, 588: True, 589: True, 590: True, 591: True, 592: True, 593: True, 594: True, 595: True, 596: True, 597: True, 598: True}
```

Reply:
but that's a form of caching 🙂 . I guess the choice is between upfront work vs lazy computation and memoisation. Even though we have a small range, during runtime an application would return maybe 1 or 2 different HTTP error codes?

Reply:
yeah, I am just thinking about maintenance costs/overhead of managing an LRU (this looks like LRU?) that requires locking vs just a static dict cache?

Reply:
I've taken special care to avoid locking when not necessary so the impact of that should be negligible (and the cache strategy should be MFU if I have implemented it right 🙂 ).